### PR TITLE
Refactor/Refactor and document the `checksum` package

### PR DESCRIPTION
**Pull request by carpawell**

### CODE REVIEW

File: checksum/doc.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+Checksum is a basic type of data checksums.
+For example, calculating checksums:
+	// retrieving any payload for hashing
+
+	var sha256Sum Checksum
+	CalculateSHA256(&sha256Sum, payload) // sha256Sum contains SHA256 hash of the payload
+
+	var tzSum Checksum
+	CalculateTillichZemor(&tzSum, payload) // tzSum contains TZ hash of the payload
```
Comment (fyrchik) [inline checksum/doc.go:12]: Do you want to try #204 here?

File: checksum/doc.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/*
+Package checksum provides primitives to work with checksums of
+NeoFS objects and storage groups.
+
+Checksum type provides functionality to specify sum type and value.
+For example, retrieving payload sum of the object:
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply checksum/doc.go]: That wasn't about adding new functions in source code but to examples here.

File: checksum/checksum.go [hunk 3]
```diff
@@ -76,72 +92,32 @@ func (c *Checksum) SetTillichZemor(v [64]byte) {
...
+func (c Checksum) String() string {
+	v2 := (refs.Checksum)(c)
+	return hex.EncodeToString(v2.GetSum())
 }
 
-func (c *Checksum) String() string {
-	return hex.EncodeToString((*refs.Checksum)(c).GetSum())
+// Empty returns true if it is called on
+// zero checksum.
+func (c Checksum) Empty() bool {
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply checksum/checksum.go:129]: We declare unset property on type carriers, not type itself.

File: checksum/checksum.go [hunk 4]
```diff
@@ -76,72 +92,32 @@ func (c *Checksum) SetTillichZemor(v [64]byte) {
...
-// UnmarshalJSON decodes Checksum from protobuf JSON format.
-func (c *Checksum) UnmarshalJSON(data []byte) error {
-	return (*refs.Checksum)(c).UnmarshalJSON(data)
+// String implements fmt.Stringer.
+//
+// String is designed to be human-readable, and its format MAY differ between
+// SDK versions.
+func (c Checksum) String() string {
+	v2 := (refs.Checksum)(c)
+	return hex.EncodeToString(v2.GetSum())
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply checksum/checksum.go]: Nvm for now, we can change it freely.

File: checksum/checksum.go [hunk 5]
```diff
@@ -55,93 +67,74 @@ func (c *Checksum) Type() Type {
...
-
-func Equal(cs1, cs2 *Checksum) bool {
-	return cs1.Type() == cs2.Type() && bytes.Equal(cs1.Sum(), cs2.Sum())
-}
-
-// Marshal marshals Checksum into a protobuf binary form.
-func (c *Checksum) Marshal() ([]byte, error) {
-	return (*refs.Checksum)(c).StableMarshal(nil)
+// See also SetSHA256.
+func CalculateSHA256(c *Checksum, v []byte) {
```
[the remark this answers is not in the document]
Comment (carpawell) [reply checksum/checksum.go]: hm, agree, added

File: checksum/checksum.go [hunk 6]
```diff
@@ -1,15 +1,23 @@
...
 )
 
-// Checksum represents v2-compatible checksum.
+// Checksum represents checksum of some digital data.
+//
+// Checksum is mutually compatible with github.com/nspcc-dev/neofs-api-go/v2/refs.Checksum
+// message. See ReadFromV2 / WriteToV2 methods.
+//
+// Instances can be created using built-in var declaration.
+//
```
[the remark this answers is not in the document]
Comment (carpawell) [reply checksum/checksum.go:18]: /cc @cthulhu-rider

File: checksum/checksum.go [hunk 7]
```diff
@@ -1,15 +1,23 @@
...
 	"fmt"
 
 	"github.com/nspcc-dev/neofs-api-go/v2/refs"
+	"github.com/nspcc-dev/tzhash/tz"
 )
 
-// Checksum represents v2-compatible checksum.
+// Checksum represents checksum of some digital data.
+//
+// Checksum is mutually compatible with github.com/nspcc-dev/neofs-api-go/v2/refs.Checksum
```
[the remark this answers is not in the document]
Comment (carpawell) [reply checksum/checksum.go:14]: do not see a big difference, but the same words were used in @cthulhu-rider's PRs

File: checksum/checksum.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply checksum/checksum.go]: i think that `Empty` is a good alternative to pointers, but @cthulhu-rider does not agree

added [set flag](https://github.com/nspcc-dev/neofs-sdk-go/pull/200/commits/3aa52112aa9020725656395b283195a305db0776)

@alexvanin, @fyrchik, thoughts?

[the remark this answers is not in the document]
Comment (alexvanin) [reply checksum/checksum.go]: > We declare unset property on type instance carriers, not type itself.

For some reason, I thought we are going to declare those are on types. Flag in the getter function is okay for me. I did it like this originally with `bearer.Issuer()` but reverted it back. I guess I can revert it one more time.

**Review by cthulhu-rider — APPROVED** [review]

**Pull request merged**